File: models/booking/BookingInternRequest.ts
import mongoose, { Schema, Document } from "mongoose";

export interface IBookingInternRequest extends Document {
  partnerId: string; // Reference to Partner
  roomName: string;
  residentEmail: string;
  category: "spa/clubs" | "restaurant";
  status: "new" | "accepted" | "completed" | "no-show" | "canceled";
  assignee?: {
    name: string;
    staffId: string;
    profilePic?: string;
  };
  reservation: {
    date: string; // e.g. "Jan 12, 2025"
    time: string; // e.g. "12:30 PM"
  };
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

const bookingInternRequestSchema = new Schema<IBookingInternRequest>(
  {
    partnerId: {
      type: String,
      required: true,
      trim: true,
    },
    roomName: { type: String, required: true },
    residentEmail: { type: String, required: true },

    category: {
      type: String,
      enum: ["spa/clubs", "restaurant"],
      required: true,
    },

    status: {
      type: String,
      enum: ["new", "accepted", "completed", "no-show", "canceled"],
      default: "new",
    },

    assignee: {
      name: String,
      staffId: String,
      profilePic: String,
    },

    reservation: {
      date: { type: String, required: true }, // e.g. "Jan 12, 2025"
      time: { type: String, required: true }, // e.g. "12:30 PM"
    },

    notes: { type: String, trim: true },
  },
  { timestamps: true } // auto adds createdAt and updatedAt
);

// Indexes for better query performance
bookingInternRequestSchema.index({ partnerId: 1 });
bookingInternRequestSchema.index({ status: 1 });
bookingInternRequestSchema.index({ createdAt: -1 });

// Avoid model overwrite error in dev
if (
  process.env.NODE_ENV !== "production" &&
  mongoose.models.BookingInternRequest
) {
  delete mongoose.models.BookingInternRequest;
}

export const BookingInternRequest = mongoose.model<IBookingInternRequest>(
  "BookingInternRequest",
  bookingInternRequestSchema
);

export default BookingInternRequest;

File: luxeli_app/lib/features/bookings/models/booking_intern_request.dart
class BookingInternRequest {
  final String id;
  final String roomName;
  final String residentEmail;
  final String category;
  final String status;
  final Assignee? assignee;
  final Reservation reservation;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  BookingInternRequest({
    required this.id,
    required this.roomName,
    required this.residentEmail,
    required this.category,
    required this.status,
    this.assignee,
    required this.reservation,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory BookingInternRequest.fromJson(Map<String, dynamic> json) {
    return BookingInternRequest(
      id: json['_id'] as String? ?? json['id'] as String,
      roomName: json['roomName'] as String? ?? '',
      residentEmail: json['residentEmail'] as String? ?? '',
      category: json['category'] as String? ?? '',
      status: json['status'] as String? ?? 'new',
      assignee: json['assignee'] != null
          ? Assignee.fromJson(json['assignee'] as Map<String, dynamic>)
          : null,
      reservation: Reservation.fromJson(
        json['reservation'] as Map<String, dynamic>,
      ),
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'roomName': roomName,
      'residentEmail': residentEmail,
      'category': category,
      'status': status,
      'assignee': assignee?.toJson(),
      'reservation': reservation.toJson(),
      'notes': notes,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class Assignee {
  final String name;
  final String staffId;
  final String? profilePic;

  Assignee({required this.name, required this.staffId, this.profilePic});

  factory Assignee.fromJson(Map<String, dynamic> json) {
    return Assignee(
      name: json['name'] as String? ?? '',
      staffId: json['staffId'] as String? ?? '',
      profilePic: json['profilePic'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {'name': name, 'staffId': staffId, 'profilePic': profilePic};
  }
}

class Reservation {
  final String date;
  final String time;

  Reservation({required this.date, required this.time});

  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
      date: json['date'] as String? ?? '',
      time: json['time'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {'date': date, 'time': time};
  }
}
